toc/rfc/rfc-draft-pcap-bosh.md

```diff
@@ -2,60 +2,224 @@
+
+Akin to the *BOSH SSH* feature, a *BOSH pcap* feature is proposed with a similar but slightly different workflow.
+The well-known `tcpdump` tool is launched on each VM via SSH and transmits its data via the SSH channel.
+In order to handle multiple VMs and merge their captures, multiple SSH sessions to the respective targets are opened in parallel, 
+merged into a single stream in the bosh-cli and available for writing to disk or an output stream.
+
+#### Architecture Overview
+
+The architecture of the pcap-lite solution looks as follows:
+
+![](rfc-draft-pcap-bosh/tcpdump-bosh-pcap-lite.drawio.svg "Architecture Overview for the pcap-lite Solution")
+
+*Figure 2 - Architecture Overview for the pcap-lite Solution*
+
+The main benefit of operator convenience over raw use of command line tools requires the following:
+* Automatic determination of VM target addresses via the BOSH Director.
+* A single SSH session can contain multiple streams, one for data, one for process control. The Go SSH library always opens such a "Master Mode" connection.
+* Multiplexing can be done in the BOSH CLI, using a similar mechanism to pcap-release, i.e. by using the `gopacket/gopacket` libraries (and `libpcap` under the hood).
+
+The largest constraint is that the capturing is done used by tools that must already be available on the target VMs, e.g. `tcpdump` that is available as part of the stemcell. That said, `tcpdump` is available on CF stemcells and can be readily invoked by BOSH users via BOSH SSH.
+
+#### Use Cases
+
+The cases from the [Use Case Overview](#use-case-overview) are addressed below:
+
+| # | Use Case                                                                             | Comment                                                                                                                                                                                                             |
+|---|--------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| 1 | Capturing Data from specific instances                                               | Instance groups and their instances can be determined by querying the BOSH Director with an appropriate token. The invocation of `tcpdump` can be constructed with appropriate command line arguments.              |
+| 2 | Providing [pcap-filter](https://www.tcpdump.org/manpages/pcap-filter.7.html) Filters | Filters are checked for length to avoid accidental overload or attacks, and provided to the invocation of `tcpdump` on the respective target VMs.                                                                   |                                                                                                                                         |
```

Comment:
Are they checked? I would say no because, as discussed previously, there isn't really a reason to do so.

Reply:
I can write 'can be checked'. You still might want some safeguards?
But I see your point. You can do the same via BOSH SSH either way.

Reply:
Removed the checking bit.

Reply:
It may be a misunderstanding that just because someone can bypass any security checks means there should be no security checks. The reason we should sanitize input is not so much to protect against hackers but to avoid accidents. Even admins make mistakes.

We should not just pass whatever user input is given straight to "sudo tcpdump -w -" as verbatim. The user should be able to provide filter input, interface, snaplength etc. but not the actual `-i`, `-s`, `-f` parameters. If we do that we have no control over what will come back, which format it may have etc. And we won't have any means to confidently merge the returned pcap stream.

Reply:
The idea is not to just pass on flags verbatim, but pass on the pcap-filter expression itself verbatim. interface, snaplen, etc. would become parameters to the `bosh pcap` subcommand.